OpenEdXMobile/src/main/java/org/edx/mobile/receivers/MediaStatusReceiver.java

```diff
@@ -0,0 +1,150 @@
+import org.edx.mobile.module.db.DataCallback;
+import org.edx.mobile.module.db.IDatabase;
+import org.edx.mobile.module.prefs.LoginPrefs;
+import org.edx.mobile.module.prefs.PrefManager;
+import org.edx.mobile.util.FileUtil;
+import org.edx.mobile.util.Sha1Util;
+
+import java.io.File;
+import java.lang.annotation.Target;
+import java.util.List;
+
+import de.greenrobot.event.EventBus;
+import roboguice.receiver.RoboBroadcastReceiver;
+
+
+/**
+ * BroadcastReceiver to receive the removable storage (such as SD-card) status events.
+ */
+
+public class MediaStatusReceiver extends RoboBroadcastReceiver {
+    @Inject
+    private IDatabase db;
+    @Inject
+    private LoginPrefs loginPrefs;
+
+    private PrefManager prefManager;
+
+    public MediaStatusReceiver() {
+    }
+
+    @Override
+    public void handleReceive(Context context, Intent intent) {
+        prefManager = new PrefManager(context, PrefManager.Pref.SD_CARD);
+        String username = loginPrefs.getUsername();
+        String hashedUsername = (username != null) ? Sha1Util.SHA1(username) : null;
+
+        // TODO: Is this Path a good method for Identifying the path in the DB
+        // should we updated the model to have an SD Card Flag?
+        final String sdCardPath = intent.getDataString().replace("file://", "");
+
+        String action = intent.getAction();
+        if (action != null) {
+            boolean sdCardAvailable = false;
+            switch (action) {
+                case Intent.ACTION_MEDIA_REMOVED:
+                case Intent.ACTION_MEDIA_UNMOUNTED:
+                    sdCardAvailable = false;
+                    handleSDCardUnmounted(hashedUsername, sdCardPath);
+                    break;
+                case Intent.ACTION_MEDIA_MOUNTED:
+                    sdCardAvailable = true;
+                    handleSDCardMounted(context, hashedUsername, sdCardPath);
+                    break;
+            }
+            EventBus.getDefault().postSticky(new MediaStatusEvent(sdCardAvailable));
+        }
+    }
+
+    private void handleSDCardUnmounted(String hashedUsername, final String sdCardPath) {
+        db.getAllVideos(hashedUsername, new DataCallback<List<VideoModel>>() {
+            @Override
+            public void onResult(List<VideoModel> result) {
+                for (VideoModel videoModel : result) {
+                    updateVideoDownloadState(
+                            videoModel,
+                            DownloadEntry.DownloadedState.ONLINE.ordinal(),
+                            sdCardPath
+                    );
+                }
+            }
+
+            @Override
+            public void onFail(Exception ex) {
+                Log.e(this.getClass().getSimpleName(),
+                        "Unable to get to get list of Videos"
+                );
+            }
+        });
+    }
+
+    private void handleSDCardMounted(final Context context, String hashedUsername, final String sdCardPath) {
```

Comment:
There is a lot of functionality is implemented in this function. It should be split up into multiple functions. The code should be properly documented.
Make common utility functions where possible.